tmt/cli.py

```diff
@@ -2081,3 +2081,100 @@ def completion_fish(context: Context, install: bool, **kwargs: Any) -> None:
     Setup shell completions for fish.
     """
     setup_completion('fish', install)
+
+
+@click.command(context_settings={
+    "ignore_unknown_options": True,
+    "allow_extra_args": True,
+    })
+@click.pass_context
+@click.option(
+    '-r', '--root', metavar='PATH', show_default=True, default='.',
+    help='Variable passed to tmt main. See `tmt --help`')
+@click.option(
+    '-v', '--verbose', count=True, default=0,
+    help='Variable passed to tmt main. See `tmt --help`')
+@click.option(
+    '-d', '--debug', count=True, default=0,
+    help='Variable passed to tmt main. See `tmt --help`')
+@click.option(
+    '--no-color', is_flag=True, default=False,
+    help='Variable passed to tmt main. See `tmt --help`'
+    )
+@click.option(
+    '--force-color', is_flag=True, default=False,
+    help='Variable passed to tmt main. See `tmt --help`'
+    )
+@click.option(
+    '--version', is_flag=True,
+    help="Variable passed to tmt main. See `tmt --help`")
+@click.option(
+    '--lint-what', default=None,
+    type=click.Choice(['tests', 'plans', 'stories']),
+    help='tmt types to lint')
+def pre_commit(
+        context: click.Context,
+        root: str,
+        verbose: int,
+        debug: int,
+        version: bool,
+        no_color: bool,
+        force_color: bool,
+        lint_type: Optional[str]) -> None:
+    """
+    Cli wrapper of tmt lint.
+    """
+    # Special handling for --version
+    if version:
+        main(['--version'])
+
+    # Create logger to reuse tmt.utils.Command utility
+    # apply_colors_output is not used because this is recreated and handled by main app instead
+    apply_colors_output, apply_colors_logging = tmt.log.decide_colorization(no_color, force_color)
+    logger = tmt.log.Logger.create(
+        apply_colors_output=apply_colors_output,
+        apply_colors_logging=apply_colors_logging
+        )
+    logger.add_console_handler()
+
+    # Check that .fmf/version file is present for the specific root requested
+    try:
+        git_command = tmt.utils.Command('git', 'rev-parse', '--show-toplevel')
+        git_root = git_command.run(cwd=None, logger=logger).stdout
+        if not git_root:
+            logger.fail("git rev-parse did not produce a path")
+            sys.exit(1)
+        git_root = git_root.strip()
+        git_command = tmt.utils.Command(
+            'git', 'ls-files', '--error-unmatch', f"{git_root}/{root}/.fmf/version"
+            )
+        git_command.run(cwd=None, logger=logger)
+    except tmt.utils.RunError as err:
+        # Forward git command error for more legible output
+        if err.stderr:
+            logger.fail(err.stderr)
```

Comment:
I'm not sure about this, why not use `tmt.utils.render_exception()`, or even `tmt.utils.show_exception()`?

Which is BTW the default handling in `tmt.__main__.run_cli()` - is there a reason to not use `run_cli()` which would then take care of pesky details like logging exceptions and raising the right exit code? Because it seems to me we're taking a shortcut, bypassing well-known tmt entrypoint setup which provides added value and can be parameterized to call either `tmt.cli.main()` or `tmt.cli.pre_commit()`. We may end up fixing a bug in the tmt entry point & then duplicating the fix to this function. It would save us some duplication, use the same code to do one thing every time a tmt script is called, one place to fix bugs related to reporting errors, and so on.

Reply:
This was to address [this comment](https://github.com/teemtee/tmt/pull/2007#issuecomment-2045140556) basically to have the output be as readable as possible without any python taceback, exception chains, etc. Basically to print out the direct stderr of the `git` command

Reply:
I see. But I disagree with the assumption this place is exceptional enough to deserve its own "exception logging".

* if the output of `show_exception()` is not readable enough to be worth being used here, or not as readable as possible, then we should address that ASAP because it's **the** code logging every single exception tmt raises. Keeping it unreadable and doing something else in the hope of providing readable output here if nowhere else would make things worse. Nothing would be fixed and we would have two ways of logging error.
* the only exception logged with `logger.fail()` would be the one caused by `git` command failing - every other exception would lead to a full traceback because there is no other `try`/`except` catching them. So `logger.fail()` does not fix the problem of tracebacks possibly exploding in the user's face, they just would not be related to `git` commands. This would be a point in favor of re-using `tmt.__main__.main()` which takes care of all of this already.
* tracebacks are not shown by default with `show_exception()`. You get the message and possibly the output of the command as well if it was a `RunError`, but not tracebacks. Yes, chains are displayed as well, are they really that bad? In the case of `RunError`, there would probably be no such chain at all when the command fails. Tracebacks can be enabled if needed, with `TMT_SHOW_TRACEBACK=1` - if we would use `show_exception()`, I could just `export` the variable and re-run `pre-commit` to get more details should I need them for investigation. With `logger.fail()`, I cannot.

All in all, to me this place does not look special enough to deserve its own exception logging or missing all the nice and fancy goods we have prepared in `tmt.__main__.main()` for presenting nice face toward users. If there are any issues with that, I'm all for fixing them.

Reply:
Should clarify, I am ok with adding `render/show_exception`, I just haven't had the opportunity to try it out to see how it works. I was just explaining the expected requirements.

But the more important thing is to discuss the comments in: https://github.com/teemtee/tmt/pull/2007#discussion_r1569397492. That would determine if this code section can use tmt native workflow, or not.